### PR TITLE
Treat &... sequences that are not entities as regular text.

### LINKED ISSUE #1667: Unrecognized entities give strange `undefined`
If you compile this JSX:

```
<CalEvent href="/events?fields=(id,summary)&key=YOUR_KEY" />
```

it turns into:

```
CalEvent( {href:"/events?fields=(id,summary)undefinedEY"} )
```

which seems subpar. I'd expect this to throw an error.

### PR DESCRIPTION
**Pull request by RReverser**
Fixes facebook/react#1667 on parser level.

### CODE REVIEW

File: esprima.js [hunk 1]
```diff
@@ -5630,8 +5630,11 @@ parseYieldExpression: true
             entity = String.fromCharCode(parseInt(str.substr(2), 16));
```
Comment (syranide) [inline esprima.js]: Perhaps we should add a stricter check before proceeding to parse `&#` and `&x`, I had implemented it in:
https://github.com/facebook/react/pull/1514/files#diff-1c729dad5915bb01dd4d59c56c34cf73R502

Comment (RReverser) [inline esprima.js]: Makes sense, let me fix that.

**Comment by RReverser** [comment]
I went even further and covered more edge cases, i.e.:
- `&#123456789` (any 10 chars after `&`) was parsed as `String.fromCharCode(123456789)`; now parsed as-is.
- `&#123*+-;` (one or more non-numeric characters in entity code) was parsed as `String.fromCharCode(123)`, ignoring the fact that entire code substring is not a number; now parsed as-is.
- `&#;`, `&#x;`, `&#abra;` (no [numeric] characters in code at all) were parsed as `String.fromCharCode(0)`; now parsed as-is.

/cc @syranide

**Comment by syranide** [comment]
@RReverser I had `/^#x[0-9a-f]{1,2}$/` which makes it only accept 1 or 2 chars (although yes, `;` should've been included there), only 1 or 2 chars is valid AFAIK. Also, to avoid octal parsing in older browsers, use `parseInt(value, 10)` instead of that complicated regex.

**Comment by RReverser** [comment]
@syranide I removed `parseInt` in favor of `Number` exactly because of edge cases where non-numeric characters are ignored in code - `parseInt` doesn't try to treat entire string as number, it only parses as long as finds numeric characters, and silently skips the rest of string. And I don't think that `/^0+/` is complicated regex :P

**Comment by RReverser** [comment]
Also, code can contain actually more than 2 digits - check out http://www.w3.org/TR/html4/charset.html#h-5.3.1 / http://www.w3.org/2000/07/8378/xhtml/entities/entities-xhtml-math.xhtml for examples.

**Comment by syranide** [comment]
@RReverser But isn't it better to use a strict regex than fallback when it's NaN? Ah you're right about there being no char limit.

```
var numericEntity = str.match(/^#(x([0-9a-f]+)|([0-9]+))$/);
if (numericEntity) {
  if (numericEntity[1][0] === 'x') {
    codeStr = parseInt(numericEntity[2], 16);
  } else {
    codeStr = parseInt(numericEntity[3], 10);
  }
} else {
  ...
```

Seems a lot cleaner and readable to me (the second if-condition can be simplified, but it makes it less clear what's going on).

**Comment by RReverser** [comment]
@syranide And now you got a complex regexp you were firstly arguing against :) (also, you forgot to add `i` flag to it as A-F should be supported as well)

Parser itself is already a state machine, and it usually doesn't make much sense to use another complex state machine with own syntax (regexp) in places where you can easily avoid it and replace with clear parsing logic. In alternative reality, people could write entire parsers using `RegExp::match` and `RegExp::replace`, but supporting it would be much of pain.

**Comment by RReverser** [comment]
@syranide But I do agree in general that `isNaN` may be also not the best possible solution - ideally, there should be logic branching (`if`) after reading first character which would parse numbers or entity names depending on `#`-condition. However, I'd rather wait for @jeffmo's feedback on all this stuff before continuing our discussions on implementation details.

**Comment by syranide** [comment]
@RReverser I never argued against using a regexp? (or? :))

Oops yes, forgot `/i`.

I kind of see what you're saying about it already being inside of a state machine (although you do use regex to strip leading zeroes... hehe). But state-machines are inherently recursive quite often (i.e, parsing of different syntax elements often uses different local state machines), so I think it's kind of a moot point (why use a large hand-written state-machine **if** you can use a small regexp state-machine, which is probably faster).

Anyway, if something can be done neatly without a regexp, that is obviously preferable, but what I'm mostly opposed to is this:

```
code = Number(codeStr);
if (!isNaN(code)) {
  return String.fromCharCode(code);
}
```

That doesn't seem very sane (and quite magic) to me. `isNaN` itself isn't so much the issue as the weird pre-processing done to accommodate `Number(...)` and that line itself. Which when I think of it, must be invalid, how would that even work for hexadecimal values? I can't seem to make it work at least. (Also, `Number('01999') === 1999` so leading zeroes doesn't seem to be an issue)

(Also yeah, I'm mostly discussing because it's interesting :))

**Comment by RReverser** [comment]
> Which when I think of it, must be invalid, how would that even work for hexadecimal values? I can't seem to make it work at least.

Why can't you? Try `Number('0x12')` in any JS engine - it should return 18 to you.

**Comment by RReverser** [comment]
> Also, Number('01999') === 1999 so leading zeroes doesn't seem to be an issue

As I said, it's about old browsers where leading zero was treated as octal marker. This is [deprecated](http://www.ecma-international.org/ecma-262/5.1/#sec-B.1) in ES5, but since there is browserified version of React, better to be sure.

**Comment by syranide** [comment]
@RReverser Ah, got confused by the code and thought the `x` was stripped. Anyway, you can replace `Number(var)` with just `+var`.

PS. IE8 doesn't treat it as octal anyway (which is usually a good indication, but perhaps some shitty android browser version does?)

**Comment by RReverser** [comment]
> Anyway, you can replace `Number(var)` with just `+var`.

Yeah, you can if you prefer this syntax (again, question of tastes :) )

**Comment by syranide** [comment]
Hehe, that's what React uses elsewhere anyway.

**Comment by RReverser** [comment]
Ok then, easy to change :)

**Comment by RReverser** [comment]
@jeffmo Ping.

**Comment by jeffmo** [comment]
Merged here: https://github.com/facebook/esprima/commit/2d60e40f3ef75104eb9431bdce7edb03facc1f5a

**Comment by syranide** [comment]
Perhaps I'm mistaken, but it seems to me as if this parses `&#-1;`, `&#0.1;` and even `&#1e1;`?
**EDIT:** I think `&#;` too? @mathiasbynens mentioned `&#Infinity;` too.

**Pull request closed without merging**